### PR TITLE
Mock out a topics endpoint with Mirage

### PR DESCRIPTION
**Pull request by pdougall1**
This should serve as an initial example of how to test drive a resource.  It uses Mirage.

### CODE REVIEW

**Comment by pdougall1** [comment]
@camthompson @gschorkopf @gschorkopf I'm just gonna merge this because it'll help me with this next bit (sign in).  I can address concerns in a subsequent PR.

File: app/controllers/topics.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import Ember from 'ember';
+
+export default Ember.Controller.extend({
+
+  newTopic: Ember.computed(function () {
+    return this.store.createRecord('topic');
+  }),
+
+  replaceNewTopic () {
+    this.set('newTopic', this.store.createRecord('topic'));
```
Comment (gschorkopf) [inline app/controllers/topics.js]: Almost feels like you should leverage `newTopic` in some way, as `this.store.createRecord('topic')` is reused.

File: app/routes/topics.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  model: function () {
+    return this.store.findAll('topic');
+  },
+
+  actions: {
+    createTopic (topic) {
+      let controller =  this.get('controller');
+      topic.save().then(function () {
+        controller.replaceNewTopic();
+      }, function () {});
```
Comment (gschorkopf) [inline app/routes/topics.js]: Is the second function (used to catch errors) needed? I think in JS if you don't use the second param it just goes away.

File: app/templates/components/topics/list-container.hbs [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<ul class="topics">
+  {{#each topics as |topic|}}
+    <li>{{topics/list-item topic=topic}}</li>
```
Comment (gschorkopf) [inline app/templates/components/topics/list-container.hbs]: Not a huge fan of naming things after HTML tags but no biggie.

File: mirage/factories/topic.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import Mirage, { faker } from 'ember-cli-mirage';
+
+export default Mirage.Factory.extend({
+  name() { return faker.name.title(); },
+  description() { return faker.lorem.sentences(); }
```
Comment (gschorkopf) [inline mirage/factories/topic.js]: Neat!

File: mirage/scenarios/default.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+export default function(/* server */) {
```
Comment (gschorkopf) [inline mirage/scenarios/default.js]: What's this file? Auto-generated? Necessary?

File: package.json [hunk 6]
```diff
@@ -33,7 +33,7 @@
     "ember-cli-htmlbars-inline-precompile": "^0.3.1",
     "ember-cli-ic-ajax": "0.2.4",
     "ember-cli-inject-live-reload": "^1.3.1",
-    "ember-cli-mirage": "0.1.11",
+    "ember-cli-mirage": "0.2.0-beta.1",
```
Comment (gschorkopf) [inline package.json]: :+1:

File: app/templates/components/topics/list-container.hbs (hunk 3 above)
Comment (pdougall1) [inline app/templates/components/topics/list-container.hbs]: ha, I wasn't even thinking it was named after li.  I was thinking it was literally an item in the list of topics.

File: tests/acceptance/topics-test.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
+import { test } from 'qunit';
+import moduleForAcceptance from 'atlantaember/tests/helpers/module-for-acceptance';
+
+const doesInclude = function (string, substrings) {
```
Comment (gschorkopf) [inline tests/acceptance/topics-test.js]: This is neat! Did you write this? There should be a library that includes nifty things like this.

File: tests/integration/components/topics/list-item-test.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+moduleForComponent('topics/list-item', 'Integration | Component | topics/list item', {
+  integration: true
+});
+
+test('displays the topic name', function(assert) {
+  let topic = { name: 'topic one' };
+  this.set('topic', topic);
+  this.render(hbs`{{topics/list-item topic=topic}}`);
+  assert.equal(this.$('.topic').text(), 'topic one\n');
```
Comment (gschorkopf) [inline tests/integration/components/topics/list-item-test.js]: The `\n` is surprising, but I guess... YavaScript??

**Comment by gschorkopf** [comment]
Looks great! Sweet test coverage.

File: mirage/scenarios/default.js (hunk 5 above)
Comment (pdougall1) [inline mirage/scenarios/default.js]: yea auto-generated.  

I don't know Mirage all that well, and this is new in 2, so I just left it.  I think it does kind of communicate that we have the version of Mirage that uses this kind of file?  I could go either way honestly.

File: tests/integration/components/topics/list-item-test.js (hunk 8 above)
Comment (pdougall1) [inline tests/integration/components/topics/list-item-test.js]: The "\n" is more about the topic name being wrapped in a block element, so the next element ends up in a new line.  That seems to get translated to a new line character when jquery asks for the text in the parent element.  

Probably should use `this.$('.topic').text().trim()` instead though.

File: app/routes/topics.js (hunk 2 above)
Comment (pdougall1) [inline app/routes/topics.js]: It's actually needed to not throw the error to the console :/  
I got it from http://alexspeller.com/server-side-validations-with-ember-data-and-ds-errors/
right under "Intro to DS.Errors"

File: tests/acceptance/topics-test.js (hunk 7 above)
Comment (pdougall1) [inline tests/acceptance/topics-test.js]: Yea I wrote it.  I was thinking it belongs in the helpers maybe?  Maybe we can evolve like a `generalHelpers` object or something?

File: app/controllers/topics.js (hunk 1 above)
Comment (pdougall1) [inline app/controllers/topics.js]: Yea I'm not really happy with this.  I don't know if it's a great idea to use the `newTopic` attribute to replace itself though.  This is the best I've got until a better thing comes along.

Comment (gschorkopf) [inline app/controllers/topics.js]: Well, probably don't use the attribute then. The combination of these two (`newTopic` and `replaceNewTopic()` doesn't really make sense. To reset `newTopic` with the exact contents of `newTopic` feels like a code smell.

File: tests/acceptance/topics-test.js (hunk 7 above)
Comment (gschorkopf) [inline tests/acceptance/topics-test.js]: Seems like a good idea.

File: tests/integration/components/topics/list-item-test.js (hunk 8 above)
Comment (gschorkopf) [inline tests/integration/components/topics/list-item-test.js]: :+1:

**Pull request merged**